File: plugins/plugin-lobster/typescript/actions/resume.ts
import {
  type Action,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  parseKeyValueXml,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/specs";
import { createLobsterService } from "../services/lobsterService";
import type { LobsterSuccessEnvelope } from "../types";

interface ResumeInput {
  token: string;
  approve: boolean;
}

function _isValidResumeInput(obj: Record<string, unknown>): boolean {
  return typeof obj.token === "string" && obj.token.length > 0 && typeof obj.approve === "boolean";
}

const EXTRACT_TEMPLATE = `Extract the approval decision from the user's message.

User message: {{text}}

The user is responding to a Lobster pipeline approval request.

Respond with XML containing:
- token: The resume token (use the provided token: {{resumeToken}})
- approve: true if user approves/confirms/says yes, false if they reject/cancel/say no

<response>
<token>{{resumeToken}}</token>
<approve>true</approve>
</response>`;

async function extractResumeInfo(
  runtime: IAgentRuntime,
  message: Memory,
  resumeToken: string
): Promise<ResumeInput | null> {
  const prompt = EXTRACT_TEMPLATE.replace("{{text}}", message.content.text ?? "").replace(
    /\{\{resumeToken\}\}/g,
    resumeToken
  );

  const result = await runtime.useModel(ModelType.TEXT_SMALL, {
    prompt,
    stopSequences: [],
  });

  logger.debug("[LobsterResume] Extract result:", result);

  const parsed = parseKeyValueXml(String(result)) as Record<string, unknown> | null;

  if (!parsed) {
    return null;
  }

  // Handle approve as string or boolean
  let approve: boolean;
  if (typeof parsed.approve === "boolean") {
    approve = parsed.approve;
  } else if (typeof parsed.approve === "string") {
    approve = parsed.approve.toLowerCase() === "true";
  } else {
    return null;
  }

  const token = parsed.token ? String(parsed.token) : resumeToken;

  if (!token) {
    return null;
  }

  return { token, approve };
}

const spec = requireActionSpec("LOBSTER_RESUME");

export const lobsterResumeAction: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    // Check if there's a pending approval token in the conversation
    const service = createLobsterService(runtime);
    const isAvailable = await service.isAvailable();
    if (!isAvailable) return false;

    // Check if the message or state contains a resume token
    const hasToken = message.content.data?.resumeToken || message.content.text?.includes("resume");
    return !!hasToken;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    stateFromTrigger: State | undefined,
    _options: HandlerOptions | undefined,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ) => {
    // Get the resume token from message data or state
    const resumeToken =
      (message.content.data?.resumeToken as string) ||
      (stateFromTrigger?.data?.resumeToken as string) ||
      "";

    if (!resumeToken) {
      if (callback) {
        await callback({
          text: "No pending approval found. Run a pipeline first that requires approval.",
          actions: ["LOBSTER_RESUME_FAILED"],
          source: message.content.source,
        });
      }
      return { success: false, text: "No resume token available" };
    }

    const resumeInfo = await extractResumeInfo(runtime, message, resumeToken);

    if (!resumeInfo) {
      if (callback) {
        await callback({
          text: "I couldn't determine your approval decision. Please say 'approve' or 'reject'.",
          actions: ["LOBSTER_RESUME_FAILED"],
          source: message.content.source,
        });
      }
      return { success: false, text: "Failed to extract approval decision" };
    }

    try {
      const service = createLobsterService(runtime);
      const envelope = await service.resume({
        token: resumeInfo.token,
        approve: resumeInfo.approve,
      });

      if (!envelope.ok) {
        if (callback) {
          await callback({
            text: `Resume failed: ${envelope.error.message}`,
            actions: ["LOBSTER_RESUME_FAILED"],
            source: message.content.source,
          });
        }
        return { success: false, text: envelope.error.message, envelope };
      }

      const successEnvelope = envelope as LobsterSuccessEnvelope;

      if (successEnvelope.status === "cancelled") {
        if (callback) {
          await callback({
            text: "Pipeline was cancelled as requested.",
            actions: ["LOBSTER_RESUME_CANCELLED"],
            source: message.content.source,
          });
        }
        return { success: true, text: "Pipeline cancelled", envelope };
      }

      if (successEnvelope.status === "needs_approval") {
        // Another approval checkpoint
        const approval = successEnvelope.requiresApproval;
        const approvalText = approval
          ? `**Another Approval Required**\n\n${approval.prompt}\n\nItems: ${JSON.stringify(approval.items, null, 2)}\n\nSay "approve" or "reject" to continue.`
          : "Pipeline requires another approval.";

        if (callback) {
          await callback({
            text: approvalText,
            actions: ["LOBSTER_NEEDS_APPROVAL"],
            source: message.content.source,
            data: {
              resumeToken: approval?.resumeToken ?? "",
              status: "needs_approval",
            },
          });
        }

        return {
          success: true,
          text: approvalText,
          envelope,
          needsApproval: true,
          resumeToken: approval?.resumeToken,
        };
      }

      const outputSummary =
        successEnvelope.output.length > 0
          ? `\n\nOutput:\n${JSON.stringify(successEnvelope.output, null, 2)}`
          : "";

      const actionText = resumeInfo.approve ? "approved and completed" : "rejected";
      const successMessage = `Pipeline ${actionText} successfully.${outputSummary}`;

      if (callback) {
        await callback({
          text: successMessage,
          actions: ["LOBSTER_RESUME_SUCCESS"],
          source: message.content.source,
        });
      }

      return { success: true, text: successMessage, envelope };
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      logger.error("[LobsterResume] Error:", errorMsg);
      if (callback) {
        await callback({
          text: `Failed to resume pipeline: ${errorMsg}`,
          actions: ["LOBSTER_RESUME_FAILED"],
          source: message.content.source,
        });
      }
      return { success: false, text: errorMsg };
    }
  },

  examples: [],
};

export default lobsterResumeAction;

File: plugins/plugin-lobster/rust/src/actions/resume.rs
//! LOBSTER_RESUME action for resuming paused pipelines

use async_trait::async_trait;
use regex::Regex;
use serde_json::Value;
use tracing::info;

use crate::generated::specs::require_action_spec;
use crate::service::LobsterService;
use crate::{Action, ActionExample, ActionResult};

/// Action to resume a paused Lobster pipeline
pub struct LobsterResumeAction {
    name: &'static str,
    description: &'static str,
    similes: Vec<&'static str>,
    examples: Vec<ActionExample>,
}

impl LobsterResumeAction {
    pub fn new() -> Self {
        let spec = require_action_spec("LOBSTER_RESUME");
        Self {
            name: spec.name,
            description: spec.description,
            similes: spec.similes.clone(),
            examples: spec
                .examples
                .iter()
                .map(|ex| ActionExample {
                    user_message: ex[0].1.to_string(),
                    agent_response: ex[1].1.to_string(),
                })
                .collect(),
        }
    }

    fn determine_approval(&self, text: &str) -> bool {
        let lower = text.to_lowercase();
        let rejection_words = ["no", "reject", "cancel", "deny", "stop", "abort"];

        for word in rejection_words {
            if lower.contains(word) {
                return false;
            }
        }
        true
    }
}

impl Default for LobsterResumeAction {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Action for LobsterResumeAction {
    fn name(&self) -> &str {
        self.name
    }

    fn similes(&self) -> Vec<&str> {
        self.similes.clone()
    }

    fn description(&self) -> &str {
        self.description
    }

    async fn validate(&self, message: &Value, state: &Value) -> bool {
        let content = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let lower = content.to_lowercase();

        // Check for pending token in state
        let pending_token = state
            .get("pendingLobsterToken")
            .and_then(|v| v.as_str());

        if pending_token.is_some() {
            let approval_words = ["approve", "yes", "continue", "reject", "no", "cancel"];
            if approval_words.iter().any(|word| lower.contains(word)) {
                return true;
            }
        }

        // Explicit resume command
        lower.contains("lobster resume")
            || (lower.contains("resume") && lower.contains("pipeline"))
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&mut LobsterService>,
    ) -> ActionResult {
        let content = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let lower = content.to_lowercase();

        // Get token from state or message
        let mut token: Option<String> = state
            .get("pendingLobsterToken")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        if token.is_none() {
            // Try to extract from message
            let re = Regex::new(r"resume\s+(\S+)").ok();
            if let Some(re) = re {
                if let Some(caps) = re.captures(&lower) {
                    token = caps.get(1).map(|m| m.as_str().to_string());
                }
            }
        }

        let token = match token {
            Some(t) => t,
            None => {
                return ActionResult {
                    success: false,
                    text: "No pending pipeline to resume. Please provide a resume token.".to_string(),
                    data: None,
                    error: Some("No token available".to_string()),
                };
            }
        };

        let approve = self.determine_approval(content);

        info!("Resuming Lobster pipeline with token: {}, approve: {}", token, approve);

        let svc = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Lobster service not available".to_string(),
                    data: None,
                    error: Some("Service unavailable".to_string()),
                };
            }
        };

        let result = svc.resume(&token, approve).await;

        if !result.success {
            return ActionResult {
                success: false,
                text: format!("Failed to resume pipeline: {}", result.error.unwrap_or_default()),
                data: None,
                error: result.error,
            };
        }

        if result.status == "needs_approval" {
            if let Some(approval) = &result.approval {
                return ActionResult {
                    success: true,
                    text: format!(
                        "Pipeline reached another approval checkpoint.\n\n**Step:** {}\n**Description:** {}\n\nReply with 'approve' or 'reject' to continue.",
                        approval.step_name, approval.description
                    ),
                    data: Some(serde_json::json!({
                        "status": "needs_approval",
                        "resumeToken": approval.resume_token,
                        "stepName": approval.step_name,
                    })),
                    error: None,
                };
            }
        }

        let action_word = if approve { "approved" } else { "rejected" };
        ActionResult {
            success: true,
            text: format!("Pipeline {} and completed successfully.", action_word),
            data: Some(serde_json::json!({
                "status": "success",
                "outputs": result.outputs,
            })),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        self.examples.clone()
    }
}
